src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+			m_midi_busy_count = 1;
+			break;
+		case 0xc/4:
+			// TODO: MIDI passthrough event
+			break;
+		case 0x18/4:
+			// TODO: MIDI update_divisor
+			break;
+	}
+
+	COMBINE_DATA(&m_midi_regs[offset]);
+}
+
+void sega_9h0_0008_state::install_game_rom()
+{
+	// Nothing (overriden by each system)
+}
+
+void sega_9h0_0008_state::machine_start()
+{
+	install_game_rom();
+
+	save_item(NAME(m_requested_fiq));
+	save_item(NAME(m_irq_wait_start_addr));
+
+	m_video_regs = make_unique_clear<uint32_t[]>(0x100/4);
+	save_pointer(NAME(m_video_regs), 0x100/4);
+	save_item(NAME(m_scale));
+
+	m_midi_regs = make_unique_clear<uint32_t[]>(0x20/4);
+	save_pointer(NAME(m_midi_regs), 0x20/4);
+	save_item(NAME(m_midi_busy_count));
+
+	save_item(NAME(m_pen_previous_input));
+	save_item(NAME(m_pen_target));
+	save_item(NAME(m_pages_previous_input));
+	save_item(NAME(m_effective_page));
+
+	m_cache_layer.allocate(1024, 512);
+}
+
+void sega_9h0_0008_state::video_start()
+{
+	m_cache_palette = std::make_unique<rgb_t[]>(0x100*SCREEN_H);
+	for (int i = 0; i < 0x100 * SCREEN_H; i++) {
+		m_cache_palette[i] = 0;
+	}
+	save_pointer(NAME(m_cache_palette.get()), 0x100*SCREEN_H);
+
+	m_tiles_layers = make_unique_clear<uint8_t[]>(0x80000);
+	m_tiles_sprites = make_unique_clear<uint8_t[]>(0x80000);
+	save_pointer(NAME(m_tiles_layers.get()), 0x80000);
+	save_pointer(NAME(m_tiles_sprites.get()), 0x80000);
+
+	m_gfxdecode->set_gfx(0, std::make_unique<gfx_element>(
+		m_palette, sega_beena_8bpp_layout, m_tiles_layers.get(), 0, 1, 0));
+	m_gfxdecode->set_gfx(1, std::make_unique<gfx_element>(
+		m_palette, sega_beena_8bpp_layout, m_tiles_sprites.get(), 0, 1, 0));
+}
+
+void sega_9h0_0008_state::machine_reset()
+{
+	m_requested_fiq = false;
+	m_irq_wait_start_addr = UNKNOWN_ADDR;
+
+	memset(m_video_regs.get(), 0, 0x100);
+	m_video_regs[0x10/4] = 0x200; // Tiles are 1x scaled by default.
+	m_scale = 1;
+	m_scanline = 0;
+
+	bool is_video_connected = m_io_video_config->read() == 0;
+	if (!is_video_connected) {
+		m_video_regs[0] |= 0x00000200;
+	}
+
+	memset(m_midi_regs.get(), 0, 0x20);
+	m_midi_busy_count = 0;
+
+	m_io_sensor_regs[0] = 7; // Pen is making contact with the target surface.
+	m_io_sensor_regs[0x2c/4] = 0xffffffff; // Pages 1 to 4 are closed.
+	m_io_sensor_regs[0x30/4] = 0x00ffffff; // Pages 5 to 6 are closed.
+	m_pen_previous_input = 0;
+	m_pen_target = 0;
+	m_pages_previous_input = 0;
+	m_effective_page = 0;
+
+	m_memcache_addr = 0;
+	m_memcache_data = 0;
+	m_memcache_i = 0;
+	m_memcache_seq = BITS_EMPTY;
+	m_memcache_state = IDLE;
+	m_memcache_state_after = {};
+
+	update_rtc();
+}
+
+void sega_9h0_0008_state::device_post_load()
+{
+	m_gfxdecode->gfx(0)->mark_all_dirty();
+	m_gfxdecode->gfx(1)->mark_all_dirty();
+}
+
+void sega_9h0_0008_state::memcache_advance(uint32_t &status)
+{
+	uint8_t event = m_memcache_events.front();
+	m_memcache_events.pop();
+
+	status = event == 6 ? 1 : 0;
+
+	switch(m_memcache_seq) {
+		case BITS_0:
```

Comment:
Please put in the space between the `switch` keyword and the opening parenthesis.